neuralmonkey/encoders/cnn_encoder.py

```diff
@@ -77,35 +96,72 @@ def image_input(self) -> tf.Tensor:
             name="input_images")
 
     @tensor
-    def image_processing_layers(self) -> List[tf.Tensor]:
+    def image_mask(self) -> tf.Tensor:
+        return tf.placeholder(
+            tf.float32,
+            shape=(None, self.image_height, self.image_width, 1),
+            name="input_mask")
+
+    def batch_norm_callback(self, layer_output: tf.Tensor) -> tf.Tensor:
+        if self.batch_normalize:
+            return tf.layers.batch_normalization(
+                layer_output, training=self.train_mode)
+        return layer_output
+
+    @tensor
+    def image_processing_layers(self) -> List[Tuple[tf.Tensor, tf.Tensor]]:
         """Do all convolutions and return the last conditional map.
 
-        Applies convolutions on the input tensor with optional max pooling.
-        All the intermediate layers are stored in the `image_processing_layers`
-        attribute.  There is not dropout between the convolutional layers, by
+        There is not dropout between the convolutional layers, by
         default the activation function is ReLU.
         """
         last_layer = self.image_input
-        image_processing_layers = []  # type: List[tf.Tensor]
+        last_mask = self.image_mask
+        last_channels = self.pixel_dim
```

Comment:
nit: neprefixujeme tyhle věci spíš `prev_`, než `last_`?

Reply:
No, ale tady je to vždcky ta _last_, tj. poslední vytvořená. (Prev a next je v těch funkcích, kde ty vrstvy vznikají, protože tam je to jasný.)

Reply:
ok